### PR TITLE
✨ Add app language selection setting

### LINKED ISSUE #678: Add language selection setting
I found no way to change app language. 

**Describe the solution you'd like**
You already have a translation platform just add a drop down menu to settings page. 

**Describe alternatives you've considered**
Idk change system language? I don't want to go that far.

### PR DESCRIPTION
**Pull request by rubenquadros**
A language selection setting is added to the settings page. The `MainActivity` now extends `AppCompatActivity` to handle changing the app language. More info on this here:
https://developer.android.com/guide/topics/resources/app-languages#androidx-impl

Closes #678

### CODE REVIEW

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt [hunk 1]
```diff
@@ -29,6 +29,24 @@ class SettingsRepository(
...
+    "no" to "Norsk",
+    "pt" to "Português",
+    "ru" to "Русский",
+    "tr" to "Türkçe",
+    "uk" to "Українська"
+  )
+
+  fun getAppLanguageCode(language: String): String {
+    return appLanguageList.filter { it.value == language }.keys.firstOrNull().orEmpty()
+  }
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt:49]: 🟡 Smells a lot like an enum class to me.

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt [hunk 2]
```diff
@@ -29,6 +32,18 @@ fun SettingsView(settingsRepository: SettingsRepository) {
...
   val currentExtendDay by settingsRepository.isDayExtended.collectAsState(false)
+  val appLanguage by settingsRepository.appLanguage.collectAsState("en")
+
+  val setAppLanguage = { language: String ->
+    val languageCode = settingsRepository.getAppLanguageCode(language)
+    if (languageCode.isNotEmpty()) {
+      settingsRepository.setAppLanguage(language) //save it to storage
+
+      val appLocale = LocaleListCompat.forLanguageTags(languageCode)
+      AppCompatDelegate.setApplicationLocales(appLocale)
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt:43]: 🟡 I believe it's not necessary to use AppCompat.

On Test files I used this:

```
private fun MainActivity.setLocale(locale: Locale) {
  val resources = baseContext.resources
  Locale.setDefault(locale)
  val config = resources.configuration
  config.setLocale(locale)
  resources.updateConfiguration(config, resources.displayMetrics)
  runOnUiThread { recreate() }
}
```

Although specific to the test case, it wasn't necessary to use AppCompat. I'm not against it, but it's a big change that I need to understand better

**Review by LeoColman — CHANGES_REQUESTED** [review]
Left a few comments.

Your code is pretty much what I had in mind! Cool :D

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt (hunk 1 above)
Comment (rubenquadros) [reply app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt:49]: Yes, this make sense!

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt (hunk 2 above)
Comment (rubenquadros) [reply app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt:43]: I can give this a try. I wanted to use the newer API and this API works with `AppCompatActivity`. You can read more about it here: https://developer.android.com/guide/topics/resources/app-languages#androidx-impl

It's your call tho on which API we should go with :)

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -30,7 +30,7 @@
         android:label="Petals"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
-        android:theme="@android:style/Theme.Holo.Light.NoActionBar">
+        android:theme="@style/AppTheme">
```
[the remark this answers is not in the document]
Comment (rubenquadros) [reply app/src/main/AndroidManifest.xml:33]: Change is related to https://github.com/LeoColman/Petals/pull/725#discussion_r1788471670

File: app/src/main/kotlin/br/com/colman/petals/MainActivity.kt [hunk 4]
```diff
@@ -56,7 +56,7 @@ import org.koin.android.ext.android.inject
 import java.time.LocalDateTime
 
 @Suppress("FunctionName")
-class MainActivity : ComponentActivity(), CoroutineScope by CoroutineScope(Dispatchers.Main) {
+class MainActivity : AppCompatActivity(), CoroutineScope by CoroutineScope(Dispatchers.Main) {
```
[the remark this answers is not in the document]
Comment (rubenquadros) [reply app/src/main/kotlin/br/com/colman/petals/MainActivity.kt:59]: Change is related to https://github.com/LeoColman/Petals/pull/725#discussion_r1788471670

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt (hunk 2 above)
Comment (rubenquadros) [reply app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt:43]: Hey @LeoColman! Did you have a chance to look at this? :)

Comment (LeoColman) [reply app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt:43]: Ok, I was thinking AppCompat was something from the past, that's what mostly caught my attetion.

So, I'm assuming it's the right way to go. We should use your way

**Review by LeoColman — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt (hunk 1 above)
Comment (rubenquadros) [reply app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt:49]: Okay, let me track a look at this and get back!

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt [hunk 5]
```diff
@@ -29,6 +32,18 @@ fun SettingsView(settingsRepository: SettingsRepository) {
     settingsRepository.decimalPrecisionList[2]
   )
   val currentExtendDay by settingsRepository.isDayExtended.collectAsState(false)
+  val appLanguage by settingsRepository.appLanguage.collectAsState("en")
+
+  val setAppLanguage = { language: String ->
+    val languageCode = settingsRepository.getAppLanguageCode(language)
+    if (languageCode.isNotEmpty()) {
+      settingsRepository.setAppLanguage(language) //save it to storage
```
[the remark this answers is not in the document]
Comment (rubenquadros) [reply app/src/main/kotlin/br/com/colman/petals/settings/SettingsView.kt]: Removed this and now we rely on the library

File: app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt (hunk 1 above)
Comment (rubenquadros) [reply app/src/main/kotlin/br/com/colman/petals/settings/SettingsRepository.kt:48]: [AppCompatDelegate.getApplicationLocales()](https://developer.android.com/reference/androidx/appcompat/app/AppCompatDelegate#getApplicationLocales()) this gives the users preferred locale which we set using `AppCompatDelegate.setApplicationLocales(appLocale)`. I'm not sure if we will get a list of supported locales

File: app/src/main/kotlin/br/com/colman/petals/settings/AppLanguages.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        "es" -> Español.name
+        "it" -> Italiano.name
+        "nl" -> Nederlands.name
+        "no" -> Norsk.name
+        "pt" -> Português.name
+        "ru" -> Русский.name
+        "tr" -> Türkçe.name
+        "uk" -> Українськ.name
+        else -> English.name
+      }
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/settings/AppLanguages.kt:30]: 🟡
```suggestion
      return entries.firstOrNull { it.code == code }?.name ?: English.name
```

File: app/src/main/kotlin/br/com/colman/petals/settings/AppLanguages.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        Français.name -> Français.code
+        Español.name -> Español.code
+        Italiano.name -> Italiano.code
+        Nederlands.name -> Nederlands.code
+        Norsk.name -> Norsk.code
+        Português.name -> Português.code
+        Русский.name -> Русский.code
+        Türkçe.name -> Türkçe.code
+        Українськ.name -> Українськ.code
+        else -> English.code
```
Comment (LeoColman) [inline app/src/main/kotlin/br/com/colman/petals/settings/AppLanguages.kt:45]:
```suggestion
      return entries.firstOrNull { it.name == languageName }?.code ?: English.code 
```

**Review by LeoColman — CHANGES_REQUESTED** [review]
Getting close.
There are a few changes to make and fix the linter complaints then I think we are good to go!

**Comment by rubenquadros** [comment]
> Getting close. There are a few changes to make and fix the linter complaints then I think we are good to go!

What do we do about `Enum entry names should match the pattern` for `AppLanguage`? 😬

**Comment by LeoColman** [comment]
I think we can go with having the English name as the enum name and a variable inside the enum having a string with the localized name.

I agree with detekt that it's a bit weird to have enums with cyrilic 😂

**Review by LeoColman — APPROVED** [review]
Great PR!
I'll test locally and merge.

**Review by LeoColman — CHANGES_REQUESTED** [review]
@rubenquadros Couldn't get it to work here :/


https://github.com/user-attachments/assets/84f79434-688e-43ac-a63f-ffae6ab321b5

**Comment by rubenquadros** [comment]
> @rubenquadros Couldn't get it to work here :/
> 
>  Screencast.from.2024-10-09.17-43-43.mp4

Can you share the device details? Could have missed some version specific changes

**Comment by LeoColman** [comment]
It's an emulator:

`Properties
avd.ini.displayname              Pixel 2 API 33
avd.ini.encoding                 UTF-8
AvdId                            Pixel_2_API_33
disk.dataPartition.size          6442450944
fastboot.chosenSnapshotFile      
fastboot.forceChosenSnapshotBoot no
fastboot.forceColdBoot           no
fastboot.forceFastBoot           yes
hw.accelerometer                 yes
hw.arc                           false
hw.audioInput                    yes
hw.battery                       yes
hw.camera.back                   virtualscene
hw.camera.front                  emulated
hw.cpu.ncore                     4
hw.device.hash2                  MD5:55acbc835978f326788ed66a5cd4c9a7
hw.device.manufacturer           Google
hw.device.name                   pixel_2
hw.dPad                          no
hw.gps                           yes
hw.gpu.enabled                   yes
hw.gpu.mode                      auto
hw.initialOrientation            Portrait
hw.keyboard                      yes
hw.lcd.density                   420
hw.lcd.height                    1920
hw.lcd.width                     1080
hw.mainKeys                      no
hw.ramSize                       1536
hw.sdCard                        yes
hw.sensors.orientation           yes
hw.sensors.proximity             yes
hw.trackBall                     no
image.androidVersion.api         33
image.sysdir.1                   system-images/android-33/google_apis/x86_64/
PlayStore.enabled                false
runtime.network.latency          none
runtime.network.speed            full
showDeviceFrame                  yes
skin.dynamic                     yes
tag.display                      Google APIs
tag.id                           google_apis
vm.heapSize                      228
`

**Comment by rubenquadros** [comment]
Yes it does not seem to be working with a fresh install😭 When I was making the changes and testing it was working. I will have to check what is happening. It would be great if you also take a look. More eyes the better! 

https://github.com/user-attachments/assets/51b6c2f7-cdfe-45db-a4c8-d7397089f5d0

**Comment by rubenquadros** [comment]
Nvm I got the issue. It was caused by yesterday's refactoring:
```
 fun getAppLanguageCode(languageName: String): String {
      return entries.firstOrNull { it.name == languageName }?.languageCode ?: English.languageCode
    }
```

Notice it? It should be `it.languageName` and not `it.name` 😝

**Review by LeoColman — APPROVED** [review]

**Comment by LeoColman** [comment]
Thanks @rubenquadros ! Good Job :D

**Pull request merged**